### ISSUE TITLE
Mask with unions

### ISSUE THREAD

**Issue by lvanhala** [ianstormtaylor/superstruct#803]
Is there a way to make `mask` work with `union`? I'd like to define only one parser that I could use to:
- Validate object and make sure it does not contain any extra fields
- Parse object ignoring extra fields

I know I could use `type()` instead of `object()` but that would essentially require defining two parsers for the whole huge object tree.

Here's the minimal use case:

```
import { object, string, mask, literal, union } from 'superstruct'

const ItemA = object({
  type: literal('a'),
  valueA: string(),
})

const ItemB = object({
  type: literal('b'),
  valueB: string(),
})

// Works
mask(
  {
    type: 'a',
    valueA: 'test',
    extraField: true,
  },
  ItemA
)

// Fails
mask(
  {
    type: 'a',
    valueA: 'test',
    extraField: true,
  },
  union([ItemA, ItemB])
)

```

**Comment by dimikot** [comment]
@ianstormtaylor I believe the behavior mentioned above is actually a bugreport. Just prepared a failing test case for it independently, and then found the current issue reported:

```
  it.only('masking succeeds validation for objects with extra props in union', () => {
    const S = union([object({ a: string() }), object({ b: string() })])
    const value = { a: '1', extraProp: 42 }
    deepStrictEqual(mask(value, S), { a: '1' })
  })
```

**Comment by arturmuller** [comment]
Resolved in 2.0
